baseplate/frameworks/queue_consumer/kombu.py

```diff
@@ -0,0 +1,192 @@
+import logging
+import queue
+import socket
+
+from typing import Callable
+from typing import Optional
+from typing import Sequence
+from typing import TYPE_CHECKING
+
+import kombu
+
+from gevent.server import StreamServer
+from kombu.mixins import ConsumerMixin
+from kombu.transport.virtual import Channel
+from pyramid.config import Configurator
+
+from baseplate import Baseplate
+from baseplate import RequestContext
+from baseplate.clients.kombu import KombuSerializer
+from baseplate.frameworks.pyramid import BaseplateConfigurator
+from baseplate.server.queue_consumer import MessageHandler
+from baseplate.server.queue_consumer import PumpWorker
+from baseplate.server.queue_consumer import QueueConsumerFactory
+from baseplate.server.wsgi import make_server as make_wsgi_server
+
+
+if TYPE_CHECKING:
+    WorkQueue = queue.Queue[kombu.Message]  # pylint: disable=unsubscriptable-object
+else:
+    WorkQueue = queue.Queue
+
+
+logger = logging.getLogger(__name__)
+
+
+Handler = Callable[[RequestContext, kombu.Message], None]
+HealthCheck = Callable[[RequestContext], bool]
+
+
+class AbortError(Exception):
+    """An error that signals that the queue process should exit.
+
+    Raising an Exception that is a subclass of AbortError will cause the
+    KombuMessageHandler to re-raise the exception rather than swallowing it
+    which will cause the handler thread/process to stop.  This, in turn, will
+    gracefully shut down the QueueConsumerServer currently running.
+
+    Exceptions of this nature should be reserved for errors that are due to
+    problems with the environment rather than the message itself.  For example,
+    a node that cannot get its AWS credentials.
+    """
+
+
+class KombuConsumerWorker(ConsumerMixin, PumpWorker):  # pylint: disable=abstract-method
+    """Consumes messages from the given queues and pumps them into the internal work_queue.
+
+    This class does not directly implement the abstract `run` command from
+    PumpWorker because the ConsumerMixin class already defines it.
+    """
+
+    def __init__(
+        self,
+        connection: kombu.Connection,
+        queues: Sequence[kombu.Queue],
+        work_queue: WorkQueue,
+        serializer: Optional[KombuSerializer] = None,
+    ):
+        self.connection = connection
+        self.queues = queues
+        self.work_queue = work_queue
+        self.serializer = serializer
+
+    def get_consumers(self, Consumer: kombu.Consumer, channel: Channel) -> Sequence[kombu.Consumer]:
+        args = dict(queues=self.queues, on_message=self.work_queue.put)
```

Comment:
what happens when `work_queue` is at its max size and `kombu.Consumer` tries to put a new message on?

Reply:
It blocks until something is removed.  When running under gevent, this is monkey-patched so you yield to the event loop.

The reason for doing this is if you don't bind the queue so the consumer has to wait on it, the consumer will inhale messages and cause the process to OOM if you have enough messages.